File: src/leetcode/python/ReverseVowelsOfAString_345.py
class ReverseVowelsOfAString_345(object):
    def reverseVowels(self, s):
        """
        :type s: str
        :rtype: str
        """
        vowels = {'a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U'}
        ls = list(s)
        i, j = 0, len(ls) - 1
        while i < j and i < len(ls) and j >= 0:
            while i < len(ls) and ls[i] not in vowels:
                i += 1
            if i >= j:
                break
            while j >= 0 and ls[j] not in vowels:
                j -= 1
            ls[i], ls[j] = ls[j], ls[i]
            i += 1
            j -= 1
        return ''.join(ls)

File: src/leetcode/java/ReverseVowelsOfAString_345.java
package leetcode.java.easy;

public class ReverseVowelsOfAString_345 {
  public String reverseVowels(String s) {
    char[] schar = s.toCharArray();
    int head = 0;
    int tail = schar.length - 1;
    char tmp;
    while (head < tail && head < schar.length && tail >= 0) {
      while (head < schar.length && !isVowel(schar[head])) head++;
      if (head > tail) break;
      while (tail >= 0 && !isVowel(schar[tail])) tail--;
      if (head < tail && tail >= 0 && head < schar.length) {
        tmp = schar[head];
        schar[head] = schar[tail];
        schar[tail] = tmp;
      }
      head++;
      tail--;
    }
    return new String(schar);
  }

  private static boolean isVowel(char check) {
    switch (Character.toLowerCase(check)) {
      case 'a':
      case 'e':
      case 'i':
      case 'o':
      case 'u': return true;
      default: return false;
    }
  }
}
